src/TimelineViewManager.cc

```diff
@@ -51,28 +55,33 @@ TimelineViewManager::messageSent(const QString &event_id, const QString &roomid,
 }
 
 void
-TimelineViewManager::sendTextMessage(const QString &msg)
+TimelineViewManager::messageSendFailed(const QString &roomid, int txn_id)
+{
+        auto view = views_[roomid];
+        view->handleFailedMessage(txn_id);
+}
+
+void
+TimelineViewManager::queueTextMessage(const QString &msg)
 {
         auto room_id = active_room_;
         auto view    = views_[room_id];
 
-        view->addUserMessage(matrix::events::MessageEventType::Text, msg, client_->transactionId());
-        client_->sendRoomMessage(matrix::events::MessageEventType::Text, room_id, msg);
+        view->addUserMessage(matrix::events::MessageEventType::Text, msg, client_->incrementTransactionId());
```

Comment:
Could you move the `MatrixClient::incrementTransactinionId` call inside the `TimelineView::addUserMessage` function?